lightning/src/ln/onion_utils.rs

```diff
@@ -1651,14 +1652,42 @@ impl LocalHTLCFailureReason {
 		}
 	}
 
+	fn get_onion_debug_field(&self) -> (&'static str, usize) {
+		match self {
+			Self::InvalidOnionVersion | Self::InvalidOnionHMAC | Self::InvalidOnionKey => {
+				("sha256_of_onion", 32)
+			},
+			Self::AmountBelowMinimum | Self::FeeInsufficient => ("htlc_msat", 8),
+			Self::IncorrectCLTVExpiry | Self::FinalIncorrectCLTVExpiry => ("cltv_expiry", 4),
+			Self::FinalIncorrectHTLCAmount => ("incoming_htlc_msat", 8),
+			Self::ChannelDisabled => ("flags", 2),
+			_ => ("", 0),
+		}
+	}
+
 	pub(super) fn is_temporary(&self) -> bool {
 		self.failure_code() & UPDATE == UPDATE
 	}
 
-	#[cfg(test)]
 	pub(super) fn is_permanent(&self) -> bool {
 		self.failure_code() & PERM == PERM
 	}
+
+	fn is_badonion(&self) -> bool {
+		self.failure_code() & BADONION == BADONION
+	}
+
+	fn is_node(&self) -> bool {
+		self.failure_code() & NODE == NODE
+	}
+
+	/// Returns true if the failure is only sent by the final recipient.
+	fn is_recipient_failure(&self) -> bool {
+		self.failure_code() == LocalHTLCFailureReason::IncorrectPaymentDetails.failure_code()
+			|| *self == LocalHTLCFailureReason::FinalIncorrectCLTVExpiry
+			|| *self == LocalHTLCFailureReason::FinalIncorrectHTLCAmount
+			|| *self == LocalHTLCFailureReason::MPPTimeout
```

Comment:
I think this is intended to only apply to errors that directly map from onion codes, but we could also include `InvalidKeysendPreimage` and `PaymentSecretRequired` here if we expect this method to be used elsewhere in the future.

Reply:
`InvalidKeysendPreimage` will already be covered by `self.failure_code() == LocalHTLCFailureReason::IncorrectPaymentDetails.failure_code()`

> PaymentSecretRequired

I think I'll update the doc a bit to note that this is only dealing with bolt 04 codes? Because we'd never get this variant as a sender, and `RequiredNodeFeature ` (which it maps to) isn't a recipient-specific error.